src/librustc_parse/parser/item.rs

```diff
@@ -1742,6 +1747,36 @@ impl<'a> Parser<'a> {
         ).emit();
     }
 
+    /// Checks if current token is one of tokens which cannot be nested like `kw::Enum`. In case
+    /// it is, we try to parse the item and report error about nested types.
+    fn recover_nested_adt_item(&mut self, keyword: Symbol) -> PResult<'a, bool> {
+        if self.token.is_keyword(kw::Enum) ||
+            self.token.is_keyword(kw::Struct) ||
+            self.token.is_keyword(kw::Union)  {
+
+            let prev_token = self.token.clone();
+            let item = self.parse_item()?;
+            if self.token == token::Comma {
```

Comment:
```suggestion
            self.eat(&token::Comma);
```

Reply:
I kind don't like that we manually eat the `token::Comma` here even though we have code which deals with that: https://github.com/rust-lang/rust/blob/master/src/librustc_parse/parser/item.rs#L1380-L1394 

But the alternative would be either to move this to function and call it:
```rust
            if !self.recover_nested_adt_item(kw::Enum)? {
                // Item already parsed, we need to skip this variant.
                check_variant_comma();
                continue
            }
            ...
            check_variant_comma();
```

or wrap the code into `if`:
```rust
            if self.recover_nested_adt_item(kw::Enum)? {
                let ident = self.parse_ident()?;
    
                let struct_def = if self.check(&token::OpenDelim(token::Brace)) {
                    // Parse a struct variant.
                    let (fields, recovered) = self.parse_record_struct_body()?;
                    VariantData::Struct(fields, recovered)
                } else if self.check(&token::OpenDelim(token::Paren)) {
                    VariantData::Tuple(
                        self.parse_tuple_struct_body()?,
                        DUMMY_NODE_ID,
                    )
                } else {
                    VariantData::Unit(DUMMY_NODE_ID)
                };
                ...
            }

            if !self.eat(&token::Comma) {
                if self.token.is_ident() && !self.token.is_reserved_ident() {
                    let sp = self.sess.source_map().next_point(self.prev_span);
                    self.struct_span_err(sp, "missing comma")
                        .span_suggestion_short(
                            sp,
                            "missing comma",
                            ",".to_owned(),
                            Applicability::MaybeIncorrect,
                        )
                        .emit();
                } else {
                    break;
                }
            }
```

Reply:
I think adding that much code as in the second snippet wouldn't be worth the added complexity. The issue with the manual eating of `token::Comma` should be fixed by using `self.parse_delim_comma_seq(...)`, but let's leave that for another time.

Reply:
Yeah, the complexity is not good either. Should I leave some note/todo comment?

Reply:
Sure, leaving a FIXME to consider `parse_delim_comma_seq` would be great.